### ISSUE TITLE
tlsh calculation on >2GB files causing OutOfArrayIndexException in java implementation

### ISSUE THREAD

**Issue by jhpoelen** [trendmicro/tlsh#84]
When calculating the ltsh of a file of size > Integer.MAX (2,147,483,647 ~ 2GB), the following exception is thrown (see also https://github.com/bio-guoda/preston/issues/61):

```
java.lang.ArrayIndexOutOfBoundsException: -3
at com.trendmicro.tlsh.TlshCreator.update(TlshCreator.java:180)
```

Root cause analysis uncovered a overflow of length variable of type int. Problem can be avoided (or rather, kicked down the road) by introducing length variables of type long. 

Happy to create a pull request with fix if you are accepting them.

**Comment by kevemueller** [comment]
The [https://github.com/kevemueller/kTLSH](https://github.com/kevemueller/kTLSH) implementation of TLSH does not suffer from this problem.

**Comment by jonjoliver** [comment]
The C++ implementation also fails
$ tlsh -f  file_2.1GB
file file_2.1GB: file too small

It says the file is too small

I will update the definition of TLSH to be for files where
50 <= file length in bytes < 2^31 = 2,147,483,648 bytes

**Comment by kevemueller** [comment]
@jonjoliver I am using these constants in my implementation:
```
static final long[] TOPVAL = {/* 0 */ 1, /* 1 */ 2, /* 2 */ 3, /* 3 */ 5, /* 4 */ 7, /* 5 */ 11, /* 6 */ 17,
            /* 7 */ 25, /* 8 */ 38, /* 9 */ 57, /* 10 */ 86, /* 11 */ 129, /* 12 */ 194, /* 13 */ 291, /* 14 */ 437,
            /* 15 */ 656, /* 16 */ 854, /* 17 */ 1110, /* 18 */ 1443, /* 19 */ 1876, /* 20 */ 2439, /* 21 */ 3171,
            /* 22 */ 3475, /* 23 */ 3823, /* 24 */ 4205, /* 25 */ 4626, /* 26 */ 5088, /* 27 */ 5597, /* 28 */ 6157,
            /* 29 */ 6772, /* 30 */ 7450, /* 31 */ 8195, /* 32 */ 9014, /* 33 */ 9916, /* 34 */ 10907, /* 35 */ 11998,
            /* 36 */ 13198, /* 37 */ 14518, /* 38 */ 15970, /* 39 */ 17567, /* 40 */ 19323, /* 41 */ 21256,
            /* 42 */ 23382, /* 43 */ 25720, /* 44 */ 28292, /* 45 */ 31121, /* 46 */ 34233, /* 47 */ 37656,
            /* 48 */ 41422, /* 49 */ 45564, /* 50 */ 50121, /* 51 */ 55133, /* 52 */ 60646, /* 53 */ 66711,
            /* 54 */ 73382, /* 55 */ 80721, /* 56 */ 88793, /* 57 */ 97672, /* 58 */ 107439, /* 59 */ 118183,
            /* 60 */ 130002, /* 61 */ 143002, /* 62 */ 157302, /* 63 */ 173032, /* 64 */ 190335, /* 65 */ 209369,
            /* 66 */ 230306, /* 67 */ 253337, /* 68 */ 278670, /* 69 */ 306538, /* 70 */ 337191, /* 71 */ 370911,
            /* 72 */ 408002, /* 73 */ 448802, /* 74 */ 493682, /* 75 */ 543050, /* 76 */ 597356, /* 77 */ 657091,
            /* 78 */ 722800, /* 79 */ 795081, /* 80 */ 874589, /* 81 */ 962048, /* 82 */ 1058252, /* 83 */ 1164078,
            /* 84 */ 1280486, /* 85 */ 1408534, /* 86 */ 1549388, /* 87 */ 1704327, /* 88 */ 1874759, /* 89 */ 2062236,
            /* 90 */ 2268459, /* 91 */ 2495305, /* 92 */ 2744836, /* 93 */ 3019320, /* 94 */ 3321252, /* 95 */ 3653374,
            /* 96 */ 4018711, /* 97 */ 4420582, /* 98 */ 4862641, /* 99 */ 5348905, /* 100 */ 5883796,
            /* 101 */ 6472176, /* 102 */ 7119394, /* 103 */ 7831333, /* 104 */ 8614467, /* 105 */ 9475909,
            /* 106 */ 10423501, /* 107 */ 11465851, /* 108 */ 12612437, /* 109 */ 13873681, /* 110 */ 15261050,
            /* 111 */ 16787154, /* 112 */ 18465870, /* 113 */ 20312458, /* 114 */ 22343706, /* 115 */ 24578077,
            /* 116 */ 27035886, /* 117 */ 29739474, /* 118 */ 32713425, /* 119 */ 35984770, /* 120 */ 39583245,
            /* 121 */ 43541573, /* 122 */ 47895730, /* 123 */ 52685306, /* 124 */ 57953837, /* 125 */ 63749221,
            /* 126 */ 70124148, /* 127 */ 77136564, /* 128 */ 84850228, /* 129 */ 93335252, /* 130 */ 102668779,
            /* 131 */ 112935659, /* 132 */ 124229227, /* 133 */ 136652151, /* 134 */ 150317384, /* 135 */ 165349128,
            /* 136 */ 181884040, /* 137 */ 200072456, /* 138 */ 220079703, /* 139 */ 242087671, /* 140 */ 266296456,
            /* 141 */ 292926096, /* 142 */ 322218735, /* 143 */ 354440623, /* 144 */ 389884688, /* 145 */ 428873168,
            /* 146 */ 471760495, /* 147 */ 518936559, /* 148 */ 570830240, /* 149 */ 627913311, /* 150 */ 690704607,
            /* 151 */ 759775136, /* 152 */ 835752671, /* 153 */ 919327967, /* 154 */ 1011260767, /* 155 */ 1112386880,
            /* 156 */ 1223623232, /* 157 */ 1345985727, /* 158 */ 1480584256, /* 159 */ 1628642751,
            /* 160 */ 1791507135, /* 161 */ 1970657856, /* 162 */ 2167723648L, /* 163 */ 2384496256L,
            /* 164 */ 2622945920L, /* 165 */ 2885240448L, /* 166 */ 3173764736L, /* 167 */ 3491141248L,
            /* 168 */ 3840255616L, /* 169 */ 4224281216L, /* 170 */ 4646709504L, /* 171 */ 5111380735L,
            /* 172 */ 5622519040L, /* 173 */ 6184770816L, /* 174 */ 6803248384L, /* 175 */ 7483572991L,
            /* 176 */ 8231930623L, /* 177 */ 9055123968L, /* 178 */ 9960636928L, /* 179 */ 10956701183L,
            /* 180 */ 12052370943L, /* 181 */ 13257608703L, /* 182 */ 14583370240L, /* 183 */ 16041708032L,
            /* 184 */ 17645878271L, /* 185 */ 19410467839L, /* 186 */ 21351515136L, /* 187 */ 23486667775L,
            /* 188 */ 25835334655L, /* 189 */ 28418870271L, /* 190 */ 31260756991L, /* 191 */ 34386835455L,
            /* 192 */ 37825517567L, /* 193 */ 41608071168L, /* 194 */ 45768882175L, /* 195 */ 50345768959L,
            /* 196 */ 55380346880L, /* 197 */ 60918384640L, /* 198 */ 67010226176L, /* 199 */ 73711251455L,
            /* 200 */ 81082380287L, /* 201 */ 89190617088L, /* 202 */ 98109681663L, /* 203 */ 107920658432L,
            /* 204 */ 118712725503L, /* 205 */ 130584006656L, /* 206 */ 143642402816L, /* 207 */ 158006648832L,
            /* 208 */ 173807329279L, /* 209 */ 191188066303L, /* 210 */ 210306867200L, /* 211 */ 231337566208L,
            /* 212 */ 254471331839L, /* 213 */ 279918460927L, /* 214 */ 307910328319L, /* 215 */ 338701369343L,
            /* 216 */ 372571521024L, /* 217 */ 409827917823L, /* 218 */ 450810724351L, /* 219 */ 495891791872L,
            /* 220 */ 545481015295L, /* 221 */ 600029102079L, /* 222 */ 660032028671L, /* 223 */ 726035300351L,
            /* 224 */ 798638833663L, /* 225 */ 878502772736L, /* 226 */ 966353059839L, /* 227 */ 1062988382207L,
            /* 228 */ 1169287217151L, /* 229 */ 1286216024063L, /* 230 */ 1414837633024L, /* 231 */ 1556321468416L,
            /* 232 */ 1711953739776L, /* 233 */ 1883149107199L, /* 234 */ 2071464050688L, /* 235 */ 2278610567167L,
            /* 236 */ 2506471636992L, /* 237 */ 2757119049728L, /* 238 */ 3032831098880L, /* 239 */ 3336114143231L,
            /* 240 */ 3669725675520L, /* 241 */ 4036698439680L, /* 242 */ 4440368349184L, /* 243 */ 4884405157887L,
            /* 244 */ 5372846014464L, /* 245 */ 5910131113984L, /* 246 */ 6501144199168L, /* 247 */ 7151258697727L,
            /* 248 */ 7866384908288L, /* 249 */ 8653023477760L, /* 250 */ 9518326480895L, /* 251 */ 10470159810560L,
            /* 252 */ 11517175529472L, /* 253 */ 12668893659136L, /* 254 */ 13935783182336L,
            /* 255 */ /* 15329425519609L */ Long.MAX_VALUE};
```
See (https://github.com/kevemueller/kTLSH/blob/master/ktlsh/src/main/java/app/keve/ktlsh/impl/TLSH.java)
IMO it is more logical to NOT wrap around at 12TiB, but to saturate the function there,i.e. anything above 13935783182336L should be assigned length 255. I achieve this by using Long.MAX_VALUE as the last element of the array.

In the specification you use & 0xFF to get into the one byte space, i.e. we should after 15329425519609 again have 0. I find this hard to justify.

**Comment by jonjoliver** [comment]
I believe that we resolved this issue in release 4.6.0 on 23/04/2021
Please reopen if we have not addressed the issue
